packages/adminrouter/extra/src/lib/cache.lua

```diff
@@ -9,7 +9,7 @@ local util = require "util"
 --
 -- Values assigned to these variable need to fufil following condition:
 --
--- CACHE_FIRST_POLL_DELAY << CACHE_EXPIRATION < CACHE_POLL_PERIOD < CACHE_MAX_AGE_SOFT_LIMIT < CACHE_MAX_AGE_HARD_LIMIT
+-- CACHE_EXPIRATION < CACHE_POLL_PERIOD < CACHE_MAX_AGE_SOFT_LIMIT < CACHE_MAX_AGE_HARD_LIMIT
```

Comment:
We're changing the initial startup behaviour by removing `CACHE_FIRST_POLL_DELAY`. I'd like to know why exactly we used to have a different initial delay.

Reply:
I remember that I knew exactly. Please get back to me in person if we want to understand that again.

Reply:
We must not have race conditions around the startup of Admin Router.

One design idea is that once Admin Router is up and running (in the moment it starts processing HTTP requests) it should _immediately_ be fully functional. That was the case before introducing the cache, and it should be the case with the cache. For all types of HTTP requests.

With the current architecture -- is that true?

When the first HTTP request that needs to have a populated cache is coming in -- do we hold it back, populate the cache, and answer it?

Reply:
@ahairape I expect the flow to be something like the following if we want to keep the semantics identical:

1. in `init_worker_by_lua_block`
1.1 Schedule a closure to run once with `ngx.timer.at(update_once, CACHE_FIRST_POLL_DELAY)` 
1.2 Schedule a closure to run every 30 seconds with `ngx.timer.every(update_once, CACHE_POLL_DELAY)`

The `update_once` closure calls the existing cache update logic. 
This logic already prevents updates from racing.

Now, why we have any `CACHE_FIRST_POLL_DELAY` at all is something I don't yet understand. Ideally the worker would perform the cache update synchronously in `init_worker_by_lua_block` during startup. 
I believe that isn't possible as the ngx API available during startup is very limited so we background that work instead.
Still, I don't understand why the first poll delay isn't 0.

Reply:
> Still, I don't understand why the first poll delay isn't 0.

As far as I remember there are no strong technical reasons.

However, I think there is an important reason in terms of maintenance and making the code understandable: I am pretty sure that I set this to a non-zero number because it should not make a difference whether this is set to 0 s, 2s, or 15 s. I am pretty sure that all these values (including 0) make the system subject to the _same_ type of race condition where "something" happens in the time window between worker process initialization (where the `timer.at()` is being called) and _actual_ execution of the callback; and that something must be explicitly planned for.

Running into the race condition is much more likely with a non-zero value. That is, I think with the current code (which plans for said race condition) we _could_ totally set this to 0 but I think it might raise wrong expectations for new people coming to this code base.

Reply:
To conclude the discussion: it looks like we can totally remove the `CACHE_FIRST_POLL_DELAY` (as done by the current version of the patch) if we make sure that we don't "accidentally" set it to 0. That is, the first timer-triggered cache update should come "late" so that the time window during which said race condition exists is large enough for us to regularly exercise the code path which handles the "http request-triggered cache update".